docs/syntax.md

```diff
@@ -0,0 +1,255 @@
+
+- `upper` denotes a Unicode code point categorized as an uppercase letter or titlecase letter (i.e., with General Category value Lt or Lu).
+
+- `lower` denotes a Unicode code point categorized as a lower-case letter (i.e., with General Category value Ll).
+
+- `alphanum` denotes either `upper` or `lower`; or a Unicode code point categorized as a modifier letter, other letter, decimal digit number, letter number, or other number (i.e., with General Category value Lt, Lu, Ll, Lm, Lo, Nd, Nl or No).
+
+- `space` denotes a Unicode code point categorized as a separator space (i.e., with General Category value Zs), or any of the control characters `'\t'`, `'\n'`, `'\r'`, `'\f'`, or `'\v'`.
+
+Interested readers may find details of Unicode character categories in [Section 4.5 of The Unicode Standard 15.1.0](https://www.unicode.org/versions/Unicode15.1.0/), and the [Unicode Character Database](https://unicode.org/ucd/).
+
+## Lexical syntax
+
+Tokens form the vocabulary of LambdaBuffers files.
+The classes of tokens are defined as follows.
+
+```text
+keyword         -> 'module' | 'sum' | 'prod' | 'record'
+                 | 'opaque' | 'class' | 'instance' | 'import' 
+                 | 'qualified' | 'as'
+modulename      -> uppercamelcase
+longmodulename  -> long modulename
+typename        -> uppercamelcase
+fieldname       -> lowercamelcase\keyword
+longtypename    -> long typename
+varname         -> lowers\keyword
+punctuation     -> '<=' | ',' | '(' | ')' | '{' | '}' 
+                 | ':' | ':-' | '=' | '|'
+classname       -> uppercamelcase
+longclassname   -> long uppercamelcase
+```
+
+where
+
+```text
+uppercamelcase -> upper { alphanum }
+lowercamelcase -> lower { alphanum }
+long           -> { uppercamelcase '.' }
+lowers         -> lower { lower }
+```
+
+Input files are broken into *tokens* which use the *maximal munch* rule i.e.,
+    at each point, the next token is the longest sequence of characters that
+    form a valid token.
+`space`s or line comments are ignored except as it separates tokens that
+    would otherwise combine into a single token.
+
+### Line comments
+A *line comment* starts with the terminal `'--'` followed by zero or more printable Unicode characters stopping at the first end of line (`'\n'` or `'\r\n'`).
+
+## Syntax of LambdaBuffers files
+A LambdaBuffers file defines a module that is a collection of data types, classes, instance clauses, and derive clauses.
+
+The overall layout of a LambdaBuffers file is:
+
+```text
+module -> 'module' longmodulename { import } { statement }
+```
+
+The file must specify the module's `longmodulename` where its `modulename` must match the file's name not including the `.lbf` extension.
+After, the file may contain a sequence of `import`s followed by a sequence of `statement`s.
+
+### Import
+Imports bring *entities* (types and classes) of other modules into scope.
+
+```text
+import     -> 'import' [ 'qualified' ] longmodulename [ 'as' longmodulename ] [ importspec ]
```

Comment:
We should stick to internal naming for these to make mental mapping simpler. Is there a 'shortmodulename'? I think it's just ModuleName, ModuleNamePart and Alias right?